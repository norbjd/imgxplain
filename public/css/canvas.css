#canvases {
  position: relative;
}

#canvases canvas {
  position: absolute;
  top: 0;
  border: solid black;
}

#draw {
  z-index: 1;
}

#preview {
  z-index: 2;
}
