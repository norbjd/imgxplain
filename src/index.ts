import { App } from "./app";

window.addEventListener("DOMContentLoaded", () => {
  if (!window.AudioContext) {
    const errorMessage = "Web Audio not supported, change your browser";
    alert(errorMessage);
    throw new Error(errorMessage);
  }

  new App().run();
});
