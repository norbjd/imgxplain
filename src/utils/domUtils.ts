class DOMUtils {
  static getDocumentWidth(): number {
    return Math.max(
      document.body.scrollWidth,
      document.documentElement.scrollWidth,
      document.body.offsetWidth,
      document.documentElement.offsetWidth,
      document.documentElement.clientWidth
    );
  }

  static getDocumentHeight(): number {
    return Math.max(
      document.body.scrollHeight,
      document.documentElement.scrollHeight,
      document.body.offsetHeight,
      document.documentElement.offsetHeight,
      document.documentElement.clientHeight
    );
  }

  static getTranscriptDiv(): HTMLDivElement {
    return document.querySelector("#transcript") as HTMLDivElement;
  }

  static getMainCanvas(): HTMLCanvasElement {
    return document.querySelector("#draw") as HTMLCanvasElement;
  }

  static getPreviewCanvas(): HTMLCanvasElement {
    return document.querySelector("#preview") as HTMLCanvasElement;
  }

  static getMainDiv(): HTMLDivElement {
    return document.querySelector("#main") as HTMLDivElement;
  }

  static getWaveformDiv(): HTMLDivElement {
    return document.querySelector("#waveform");
  }

  static getEditorDiv(): HTMLDivElement {
    return document.querySelector("#editor");
  }

  static getEditForm(): HTMLFormElement {
    return document.forms.namedItem("edit");
  }

  static getLoadForm(): HTMLFormElement {
    return document.forms.namedItem("load");
  }

  static getAvailableMaxWidthForMainDiv(): number {
    return (
      (this.getDocumentWidth() -
        this.getMainDivMarginLeft() -
        this.getEditorDivWidthInPx()) *
      0.95
    );
  }

  static getAvailableMaxHeightForMainDiv(): number {
    return this.getDocumentHeight() - 60;
  }

  static getEditorDivWidthInPx(): number {
    const editorDivWidth = this.getEditorDiv()
      .style.width.replace("%", "")
      .replace("px", "");
    if (editorDivWidth == "0" || editorDivWidth == "") {
      return 0;
    }
    return (this.getDocumentWidth() * parseInt(editorDivWidth)) / 100 + 20;
  }

  static getMainDivMarginLeft(): number {
    return Number(this.getMainDiv().style.marginLeft.replace("px", ""));
  }

  static setHTMLInputElementValue(
    elementName: string,
    value: string | number
  ): void {
    (document.getElementById(elementName) as HTMLInputElement).value =
      "" + value;
  }

  static getPlayPauseButton(): HTMLButtonElement {
    return document.querySelector("[data-action='play']") as HTMLButtonElement;
  }

  static getTranscriptButton(): HTMLInputElement {
    return document.querySelector("#transcript_button");
  }

  static getCCButton(): HTMLInputElement {
    return document.querySelector("#cc");
  }

  static getWaveformZoomSlider(): HTMLInputElement {
    return document.querySelector("[data-action='zoom']");
  }

  static getExportActionsToJsonButton(): HTMLButtonElement {
    return document.querySelector(
      "[data-action='export']"
    ) as HTMLButtonElement;
  }

  static getExportToVideoButton(): HTMLButtonElement {
    return document.querySelector(
      "[data-action='export-video']"
    ) as HTMLButtonElement;
  }

  static getToggleEditorButton(): HTMLButtonElement {
    return document.getElementById("toggle_editor_button") as HTMLButtonElement;
  }

  static openEditor(): void {
    this.getToggleEditorButton().textContent = "ðŸ–‰ Close Editor";
    DOMUtils.getEditorDiv().style.width = "40%";
    DOMUtils.getEditorDiv().style.borderLeft = "1px solid black";
    DOMUtils.getEditorDiv().style.paddingLeft = "10px";
    DOMUtils.getEditorDiv().style.paddingRight = "10px";
  }

  static closeEditor(): void {
    this.getToggleEditorButton().textContent = "ðŸ–‰ Open Editor";
    DOMUtils.getEditorDiv().style.width = "0";
    DOMUtils.getEditorDiv().style.borderLeft = "";
    DOMUtils.getEditorDiv().style.paddingLeft = "0";
    DOMUtils.getEditorDiv().style.paddingRight = "0";
  }

  static removeLoadModalDialog(): void {
    document.getElementById("load_modal_dialog").remove();
  }
}

export { DOMUtils };
