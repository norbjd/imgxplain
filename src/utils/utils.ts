class Utils {
  static isEquivalent(a: unknown, b: unknown): boolean {
    if (typeof a != typeof b) {
      return false;
    }

    const aProps = Object.getOwnPropertyNames(a);
    const bProps = Object.getOwnPropertyNames(b);

    if (aProps.length != bProps.length) {
      return false;
    }

    for (let i = 0; i < aProps.length; i++) {
      const propName = aProps[i];
      if (a[propName] !== b[propName]) {
        return false;
      }
    }

    return true;
  }
}

export { Utils };
