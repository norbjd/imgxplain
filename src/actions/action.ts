import { Canvas } from "../canvas";

class Action<T extends ActionOptions> {
  canvas: Canvas;
  type: string;
  options: T;

  constructor(canvas: Canvas, type: string, options: T) {
    this.canvas = canvas;
    this.type = type;
    this.options = options;
  }

  do(context: CanvasRenderingContext2D): void {
    throw new Error("Not implemented");
  }

  execute(): void { this.do(this.canvas.mainContext); }
  preview(): void { this.do(this.canvas.previewContext); }
}

class ActionOptions {}

export {
  Action,
  ActionOptions,
};
