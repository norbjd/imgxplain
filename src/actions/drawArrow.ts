import { Action, ActionOptions } from "./action";
import { Canvas } from "../canvas";

class DrawArrowOptions extends ActionOptions {
  originX: number;
  originY: number;
  destinationX: number;
  destinationY: number;
  lineWidth: number;
  color: string;

  constructor(
    originX: number,
    originY: number,
    destinationX: number,
    destinationY: number,
    lineWidth: number,
    color: string
  ) {
    super();
    this.originX = originX;
    this.originY = originY;
    this.destinationX = destinationX;
    this.destinationY = destinationY;
    this.lineWidth = lineWidth;
    this.color = color;
  }
}

class DrawArrow extends Action<DrawArrowOptions> {
  constructor(canvas: Canvas, options: DrawArrowOptions) {
    super(canvas, "draw_arrow", options);
  }

  do(context: CanvasRenderingContext2D): void {
    const lineCanvasOrigin = this.canvas.getPointProjectionOnCanvas(
      this.options.originX,
      this.options.originY
    );
    const lineCanvasOriginX = lineCanvasOrigin.x;
    const lineCanvasOriginY = lineCanvasOrigin.y;

    const lineCanvasDestination = this.canvas.getPointProjectionOnCanvas(
      this.options.destinationX,
      this.options.destinationY
    );
    const lineCanvasDestinationX = lineCanvasDestination.x;
    const lineCanvasDestinationY = lineCanvasDestination.y;

    context.save();

    context.lineWidth = this.options.lineWidth;
    context.strokeStyle = this.options.color;

    context.beginPath();
    context.moveTo(lineCanvasOriginX, lineCanvasOriginY);
    context.lineTo(lineCanvasDestinationX, lineCanvasDestinationY);
    context.stroke();
    context.closePath();

    context.beginPath();

    const arrowHeadRadius = this.options.lineWidth * 2;

    let angle = Math.atan2(
      lineCanvasDestinationY - lineCanvasOriginY,
      lineCanvasDestinationX - lineCanvasOriginX
    );
    context.moveTo(
      arrowHeadRadius * Math.cos(angle) + lineCanvasDestinationX,
      arrowHeadRadius * Math.sin(angle) + lineCanvasDestinationY
    );
    angle += (1.0 / 3.0) * (2 * Math.PI);
    context.lineTo(
      arrowHeadRadius * Math.cos(angle) + lineCanvasDestinationX,
      arrowHeadRadius * Math.sin(angle) + lineCanvasDestinationY
    );
    angle += (1.0 / 3.0) * (2 * Math.PI);
    context.lineTo(
      arrowHeadRadius * Math.cos(angle) + lineCanvasDestinationX,
      arrowHeadRadius * Math.sin(angle) + lineCanvasDestinationY
    );
    context.closePath();

    context.fillStyle = this.options.color;
    context.fill();

    context.restore();
  }
}

export { DrawArrow, DrawArrowOptions };
