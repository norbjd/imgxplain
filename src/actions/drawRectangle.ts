import { Action, ActionOptions } from "./action";
import { Canvas } from "../canvas";

class DrawRectangleOptions extends ActionOptions {
    topLeftX: number;
    topLeftY: number;
    bottomRightX: number;
    bottomRightY: number;
    borderSize: number;
    color: string;
  
    constructor(
      topLeftX: number,
      topLeftY: number,
      bottomRightX: number,
      bottomRightY: number,
      borderSize: number,
      color: string
    ) {
      super();
      this.topLeftX = topLeftX;
      this.topLeftY = topLeftY;
      this.bottomRightX = bottomRightX;
      this.bottomRightY = bottomRightY;
      this.borderSize = borderSize;
      this.color = color;
    }
  }
  
  class DrawRectangle extends Action<DrawRectangleOptions> {
    constructor(canvas: Canvas, options: DrawRectangleOptions) {
      super(canvas, "draw_rectangle", options);
    }
  
    do(context: CanvasRenderingContext2D): void {
      const rectCanvasTopLeft = this.canvas.getPointProjectionOnCanvas(
        this.options.topLeftX,
        this.options.topLeftY
      );
      const rectCanvasTopLeftX = rectCanvasTopLeft.x;
      const rectCanvasTopLeftY = rectCanvasTopLeft.y;
  
      const rectCanvasBottomRight = this.canvas.getPointProjectionOnCanvas(
        this.options.bottomRightX,
        this.options.bottomRightY
      );
      const rectCanvasBottomRightX = rectCanvasBottomRight.x;
      const rectCanvasBottomRightY = rectCanvasBottomRight.y;
  
      const rectCanvasWidth = rectCanvasBottomRightX - rectCanvasTopLeftX;
      const rectCanvasHeight = rectCanvasBottomRightY - rectCanvasTopLeftY;
  
      context.save();
  
      context.beginPath();
      context.lineWidth = this.options.borderSize;
      context.strokeStyle = this.options.color;
      context.rect(
        rectCanvasTopLeftX,
        rectCanvasTopLeftY,
        rectCanvasWidth,
        rectCanvasHeight
      );
      context.stroke();
  
      context.restore();
    }
  }

  export {
    DrawRectangle,
    DrawRectangleOptions,
  };