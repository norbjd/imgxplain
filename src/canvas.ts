import { Action, ActionOptions } from "./actions/action";
import { Zoom, ZoomOptions } from "./actions/zoom";
import { DOMUtils } from "./utils/domUtils";

class Canvas {
  img: HTMLImageElement;
  imgTopLeftX: number;
  imgTopLeftY: number;
  imgBottomRightX: number;
  imgBottomRightY: number;

  mainCanvas: HTMLCanvasElement;
  mainContext: CanvasRenderingContext2D;

  previewCanvas: HTMLCanvasElement;
  previewContext: CanvasRenderingContext2D;

  clickPosition: { x: number; y: number };

  ccEnabled: boolean;

  constructor(imgBlob: string) {
    this.img = new Image();
    this.img.src = imgBlob;
    this.img.addEventListener("load", () => {
      this.renderImage();
    });

    this.imgTopLeftX = 0;
    this.imgTopLeftY = 0;
    this.imgBottomRightX = this.img.naturalWidth;
    this.imgBottomRightY = this.img.naturalHeight;

    this.mainCanvas = DOMUtils.getMainCanvas();
    this.mainContext = this.mainCanvas.getContext("2d");

    this.previewCanvas = DOMUtils.getPreviewCanvas();
    this.previewContext = this.previewCanvas.getContext("2d");

    this.clickPosition = { x: 0, y: 0 };

    this.ccEnabled = false;
  }

  render(
    topLeftX: number,
    topLeftY: number,
    bottomRightX: number,
    bottomRightY: number
  ): void {
    this.imgTopLeftX = topLeftX;
    this.imgTopLeftY = topLeftY;
    this.imgBottomRightX = bottomRightX;
    this.imgBottomRightY = bottomRightY;

    const imgWidth = this.imgBottomRightX - this.imgTopLeftX;
    const imgHeight = this.imgBottomRightY - this.imgTopLeftY;

    const imgRatio = imgWidth / imgHeight;

    const mainDivMaxWidth = DOMUtils.getAvailableMaxWidthForMainDiv();
    const mainDivMaxHeight = DOMUtils.getAvailableMaxHeightForMainDiv();

    let canvasWidth = Math.min(mainDivMaxWidth, imgWidth);
    let canvasHeight = canvasWidth / imgRatio;

    if (canvasHeight >= mainDivMaxHeight) {
      canvasHeight = mainDivMaxHeight;
      canvasWidth = canvasHeight * imgRatio;
    }

    this.mainCanvas.width = canvasWidth;
    this.mainCanvas.height = canvasHeight;
    this.mainContext.drawImage(
      this.img,
      this.imgTopLeftX,
      this.imgTopLeftY,
      imgWidth,
      imgHeight,
      0,
      0,
      canvasWidth,
      canvasHeight
    );

    this.previewCanvas.width = canvasWidth;
    this.previewCanvas.height = canvasHeight;
    this.previewCanvas.style.display = "none";

    // TODO : move somewhere else
    // -20 : padding
    DOMUtils.getTranscriptDiv().style.width = "" + (canvasWidth + 5 - 20);
  }

  renderFullImage(): void {
    this.render(0, 0, this.img.naturalWidth, this.img.naturalHeight);
  }

  renderImage(): void {
    this.render(
      this.imgTopLeftX,
      this.imgTopLeftY,
      this.imgBottomRightX,
      this.imgBottomRightY
    );
  }

  getWidthAndHeight(): { width: number; height: number } {
    return {
      width: this.mainCanvas.width,
      height: this.mainCanvas.height,
    };
  }

  getImageRatio(): { ratioWidth: number; ratioHeight: number } {
    const imgWidth = this.imgBottomRightX - this.imgTopLeftX;
    const imgHeight = this.imgBottomRightY - this.imgTopLeftY;

    const canvasWidthAndHeight = this.getWidthAndHeight();

    return {
      ratioWidth: imgWidth / canvasWidthAndHeight.width,
      ratioHeight: imgHeight / canvasWidthAndHeight.height,
    };
  }

  getPointProjectionOnCanvas(
    imgPointX: number,
    imgPointY: number
  ): { x: number; y: number } {
    const ratio = this.getImageRatio();

    const canvasPointX = (imgPointX - this.imgTopLeftX) / ratio.ratioWidth;
    const canvasPointY = (imgPointY - this.imgTopLeftY) / ratio.ratioHeight;

    return {
      x: canvasPointX,
      y: canvasPointY,
    };
  }

  getCursorPositionOnCanvas(event: MouseEvent): { x: number; y: number } {
    const rect = this.mainCanvas.getBoundingClientRect();
    const x = event.clientX - rect.left;
    const y = event.clientY - rect.top;

    const ratio = this.getImageRatio();

    return {
      x: Math.floor(x * ratio.ratioWidth) + this.imgTopLeftX,
      y: Math.floor(y * ratio.ratioHeight) + this.imgTopLeftY,
    };
  }

  addClickEventListener(listener: (event: MouseEvent) => void): void {
    this.mainCanvas.addEventListener("click", listener);
  }

  executeAction(action: Action<ActionOptions>): void {
    action.execute();
  }

  executeActions(actions: Action<ActionOptions>[]): void {
    if (actions.find((action) => action instanceof Zoom) == undefined) {
      this.renderFullImage();
    }

    this.renderImage();
    actions.forEach((action) => {
      this.executeAction(action);
    });
  }

  previewAction(action: Action<ActionOptions>, timeInMs: number): void {
    this.previewCanvas.style.display = "block";
    action.preview();
    setTimeout(() => {
      this.previewContext.clearRect(
        0,
        0,
        this.previewCanvas.width,
        this.previewCanvas.height
      );
      this.previewCanvas.style.display = "none";
      if (action instanceof Zoom) {
        const unzoom = new Zoom(
          this,
          new ZoomOptions(0, 0, this.img.naturalWidth, this.img.naturalHeight)
        );
        this.executeAction(unzoom);
      }
    }, timeInMs);
  }
}

export { Canvas };
