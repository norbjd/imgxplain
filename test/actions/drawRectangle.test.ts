import { expect, it, jest, beforeEach } from '@jest/globals';
import { DrawRectangle, DrawRectangleOptions } from "../../src/actions/drawRectangle";
import { Canvas } from '../../src/canvas';
import { DOMUtils } from '../../src/utils/domUtils';

import { readFileSync } from "fs";
import { resolve } from "path";
import { initCanvas } from './utils';

const html = readFileSync(resolve(__dirname, '../../public/index.html'), 'utf8');
jest.dontMock('fs');
document.documentElement.innerHTML = html.toString();

let canvas: Canvas;
beforeEach(() => {
    canvas = initCanvas();
});

it("should draw a red rectangle", () => {
    new DrawRectangle(canvas, new DrawRectangleOptions(
        10, 20, 50, 100, 5, "#ff0000"
    )).execute();
    const imgBase64 = DOMUtils.getMainCanvas().toDataURL("image/png");
    expect(imgBase64).toMatchSnapshot();
});

it("should draw a blue rectangle", () => {
    new DrawRectangle(canvas, new DrawRectangleOptions(
        10, 40, 250, 130, 3, "#0000ff"
    )).execute();
    const imgBase64 = DOMUtils.getMainCanvas().toDataURL("image/png");
    expect(imgBase64).toMatchSnapshot();
});

it("should preview a red rectangle", () => {
    new DrawRectangle(canvas, new DrawRectangleOptions(
        10, 20, 50, 100, 5, "#ff0000"
    )).preview();
    const imgBase64 = DOMUtils.getPreviewCanvas().toDataURL("image/png");
    expect(imgBase64).toMatchSnapshot();
});

it("should preview a blue rectangle", () => {
    new DrawRectangle(canvas, new DrawRectangleOptions(
        10, 40, 250, 130, 3, "#0000ff"
    )).preview();
    const imgBase64 = DOMUtils.getPreviewCanvas().toDataURL("image/png");
    expect(imgBase64).toMatchSnapshot();
});
