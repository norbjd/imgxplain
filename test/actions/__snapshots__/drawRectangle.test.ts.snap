// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should draw a blue rectangle 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAACZklEQVR4nO3XMY6CQACGUdx4Lzka3gxONlvYbKcmIPnc95LpKP6GLzOXMcaYAAJ+zh4A8CrBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgIzrsw/W9XGAtnl+nLKXgnW/f2IKcLR6sDwJgYynN6y/brd+oeE/Wddp2razV+znrWDN8zQty1FTgL0ty3cFy5MQyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyLi+8/G6TtOyHDUF2Nu6nr1gX28Fa9seB+AMnoRAxtMb1jx/YgZwtG/4ly9jjHH2CIBXeBICGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARm/7I0gOfAXuSYAAAAASUVORK5CYII="`;

exports[`should draw a red rectangle 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAACaUlEQVR4nO3YQWrCQACG0Vi8lzlavNnkZOnCbbEqGvvZ92BWIfCvPpg5bNu2TQABX+8eAHArwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwg43j16xiXUzLPlwN8nN+DdT7vNOWJBAs+kishkCFYQIZgARnX37B+cjr9nTeiMaZpXd+9AtjJ/cGa52lalhdMecCyCBb8I66EQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQMbx7j/GmKZlecGUB4zx7gXAju4P1rpeDsDOXAmBDMECMgQLyLj+hjXPO814ouJm4CaHbdu2d48AuIUrIZAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQMY3JDQgObUhoWoAAAAASUVORK5CYII="`;

exports[`should preview a blue rectangle 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAACZklEQVR4nO3XMY6CQACGUdx4Lzka3gxONlvYbKcmIPnc95LpKP6GLzOXMcaYAAJ+zh4A8CrBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgIzrsw/W9XGAtnl+nLKXgnW/f2IKcLR6sDwJgYynN6y/brd+oeE/Wddp2razV+znrWDN8zQty1FTgL0ty3cFy5MQyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwgQ7CADMECMgQLyLi+8/G6TtOyHDUF2Nu6nr1gX28Fa9seB+AMnoRAxtMb1jx/YgZwtG/4ly9jjHH2CIBXeBICGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARm/7I0gOfAXuSYAAAAASUVORK5CYII="`;

exports[`should preview a red rectangle 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAACaUlEQVR4nO3YQWrCQACG0Vi8lzlavNnkZOnCbbEqGvvZ92BWIfCvPpg5bNu2TQABX+8eAHArwQIyBAvIECwgQ7CADMECMgQLyBAsIEOwgAzBAjIEC8gQLCBDsIAMwQIyBAvIECwg43j16xiXUzLPlwN8nN+DdT7vNOWJBAs+kishkCFYQIZgARnX37B+cjr9nTeiMaZpXd+9AtjJ/cGa52lalhdMecCyCBb8I66EQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQMbx7j/GmKZlecGUB4zx7gXAju4P1rpeDsDOXAmBDMECMgQLyLj+hjXPO814ouJm4CaHbdu2d48AuIUrIZAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQIZgARmCBWQIFpAhWECGYAEZggVkCBaQIVhAhmABGYIFZAgWkCFYQMY3JDQgObUhoWoAAAAASUVORK5CYII="`;
